core/execution/spot/holding_account_tracker.go

```diff
@@ -0,0 +1,259 @@
+	return &HoldingAccountTracker{
+		orderIDToQuantity: map[string]*num.Uint{},
+		orderIDToFee:      map[string]*num.Uint{},
+		collateral:        collateral,
+		log:               log,
+		snapshot: &types.PayloadHoldingAccountTracker{
+			HoldingAccountTracker: &types.HoldingAccountTracker{
+				MarketID: marketID,
+			},
+		},
+	}
+}
+
+func (hat *HoldingAccountTracker) GetCurrentHolding(orderID string) (*num.Uint, *num.Uint) {
+	qty := num.UintZero()
+	fees := num.UintZero()
+	if q, ok := hat.orderIDToQuantity[orderID]; ok {
+		qty = q
+	}
+	if f, ok := hat.orderIDToFee[orderID]; ok {
+		fees = f
+	}
+	return qty, fees
+}
+
+func (hat *HoldingAccountTracker) TransferToHoldingAccount(ctx context.Context, orderID, party, asset string, quantity *num.Uint, fee *num.Uint) (*types.LedgerMovement, error) {
+	if _, ok := hat.orderIDToQuantity[orderID]; ok {
+		return nil, fmt.Errorf("funds for the order have already been transferred to the holding account")
+	}
+	total := num.Sum(quantity, fee)
+
+	transfer := &types.Transfer{
+		Owner: party,
+		Amount: &types.FinancialAmount{
+			Asset:  asset,
+			Amount: total,
+		},
+		Type: types.TransferTypeHoldingAccount,
+	}
+	le, err := hat.collateral.TransferToHoldingAccount(ctx, transfer)
+	if err != nil {
+		return nil, err
+	}
+	hat.orderIDToQuantity[orderID] = quantity
+	if !fee.IsZero() {
+		hat.orderIDToFee[orderID] = fee
+	}
+	return le, nil
+}
+
+func (hat *HoldingAccountTracker) TransferFeeToHoldingAccount(ctx context.Context, orderID, party, asset string, feeQuantity *num.Uint) (*types.LedgerMovement, error) {
+	if feeQuantity.IsZero() {
+		return nil, nil
+	}
+	transfer := &types.Transfer{
+		Owner: party,
+		Amount: &types.FinancialAmount{
+			Asset:  asset,
+			Amount: feeQuantity.Clone(),
+		},
+		Type: types.TransferTypeHoldingAccount,
+	}
+	le, err := hat.collateral.TransferToHoldingAccount(ctx, transfer)
+	if err != nil {
+		return nil, err
+	}
+	hat.orderIDToFee[orderID] = feeQuantity
+	return le, nil
+}
+
+func (hat *HoldingAccountTracker) ReleaseFeeFromHoldingAccount(ctx context.Context, orderID, party, asset string) (*types.LedgerMovement, error) {
+	feeQuantity, ok := hat.orderIDToFee[orderID]
+	if !ok {
+		return nil, fmt.Errorf("failed to find locked fee amount for order id %s", orderID)
+	}
+	transfer := &types.Transfer{
+		Owner: party,
+		Amount: &types.FinancialAmount{
+			Asset:  asset,
+			Amount: feeQuantity.Clone(),
+		},
+		Type: types.TransferTypeHoldingAccount,
+	}
+	delete(hat.orderIDToFee, orderID)
+	le, err := hat.collateral.ReleaseFromHoldingAccount(ctx, transfer)
+	if err != nil {
+		return nil, err
+	}
+	return le, err
+}
+
+func (hat *HoldingAccountTracker) ReleaseQuantityHoldingAccount(ctx context.Context, orderID, party, asset string, quantity *num.Uint, fee *num.Uint) (*types.LedgerMovement, error) {
+	total := num.Sum(quantity, fee)
+	if fee != nil && !fee.IsZero() {
```

Comment:
if `fee` is allowed to be `nil` like this line suggests, we should probably also check it before doing `total := num.Sum()` above.

Reply:
I'll remove the nil check for now and verify that it can't actually be.